Step defs with parameters

jowers opened the pull request:
Here's my first attempt at allowing parameters in a step definition...

mattwynne commented on features/define_a_step.feature:
```diff
@@ -22,3 +22,27 @@ Feature: Define a step
       Hello world
       """
+
+  Scenario: Define a passing step with a parameter
```
Is it relevant that this is a *passing* step?

mattwynne commented on lib/cucumber/tcl/framework.tcl:
```diff
@@ -56,11 +65,14 @@ proc _search_steps {step_name {execute 0}} {
-    if {[regexp $existing_step_name $step_name matchresult]} {
+    if {[regexp $existing_step_name $step_name matchresult {*}[join $existing_step_params]]} {
       if {$execute == 1} {
+#	puts "JOWERS -> $existing_step_name, $step_name, $matchresult"
```
Did you mean to leave this in? ;)

jowers replied:
No! I really need a pre-commit hook that blocks checkin if it finds JOWERS anywhere in the code. I've removed that and will push up the change shortly.

mattwynne commented on lib/cucumber/tcl/framework.tcl:
```diff
@@ -12,31 +12,42 @@ set STEPS [list]
+    set re [lindex $args 0]
+    set params [lindex $args 1]
+    set body [lindex $args 2]
+  }
```
I don't pretend to understand this code completely, but I get the impression it would only work for max 3 parameters, is that right? Can you see a way to make it more generic?

jowers replied:
The way I've written this was that you can either pass in 2 parameters (i.e. a regexp and a body) or 3 parameters (i.e. regexp, list of parameters and a body), so in the case of a regexp that can capture multiple matches, the Given call would look something like:

    Given {^Match1 (\w*), Match2 (\d*), Match 3 (\w*)} {match1 match2 match3} {
      do_something
    }

Do you think it would be better to have a separate parameter for each match? Or does this look ok (maybe with an error raised if the number of parameters is incorrect?)

I've just realised that I need to updatethe README to reflect this, but will hold off in case we decide to change this.

mattwynne replied:
I get it now. No, I think it's fine.

jowers commented:
I've just checked in a couple of additional changes:
* Renamed the new scenario as suggested above
* Added a unit test to cover the case that _add_steps is called with more than 3 parameters
* Updated _add_steps to raise an error if it is called with more than 3 parameters
* Updated the README to include examples with variables to match a regexp

Anything else you think is needed for this one?

mattwynne commented:
This looks good to me @jowers. Want to merge it into master with `--squash` to keep the history nice and clean?

jowers commented:
Done (I think) had to resolve conflicts in the README, but seems to have gone through.

mattwynne commented:
Great!

Little bit of GitHub coaching: When you use `--squash` you have to explicitly close the PR issue (GH does it automatically if you merge the whole PR branch without squashing). You can either do that by hand on the GitHub UI, or by mentioning `closes #<PR or issue number>` in your commit message for the squashed merge commit.

The alternative to this is to rebase the PR branch until it's neat and tidy (using `git rebase -i` and `git push --force`) then you can just merge it in normally.

mattwynne commented:
It's also good practice to delete the PR branch once it's merged in to keep the repo tidy. I normally forget to do this - see how many dead branches there are in cucumber/cucumber, for example.
